x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java

```diff
@@ -239,18 +285,25 @@ public void onResponse(MultiGetResponse multiGetItemResponses) {
             for (MultiGetItemResponse trace : multiGetItemResponses) {
                 if (trace.isFailed() == false && trace.getResponse().isExists()) {
                     String id = trace.getId();
-                    StackTrace stacktrace = StackTrace.fromSource(trace.getResponse().getSource());
-                    stackTracePerId.put(id, stacktrace);
-                    totalFrames.addAndGet(stacktrace.frameIds.size());
-                    stackFrameIds.addAll(stacktrace.frameIds);
-                    executableIds.addAll(stacktrace.fileIds);
+                    // Duplicates are expected as we query multiple indices - do a quick pre-check before we deserialize a response
+                    if (stackTracePerId.containsKey(id) == false) {
+                        StackTrace stacktrace = StackTrace.fromSource(trace.getResponse().getSource());
+                        // Guard against concurrent access and ensure we only handle each item once
+                        if (stackTracePerId.putIfAbsent(id, stacktrace) == null) {
+                            stackTracePerId.put(id, stacktrace);
```

Comment:
Isn't this redundant? Theoretically, `stackTracePerId.putIfAbsent(id, stacktrace) == null` should have already inserted this, right?